### ISSUE TITLE
Error in wildcard generation (?)

### ISSUE THREAD

**Issue by louiejtaylor** [sunbeam-labs/sunbeam#185]
Context: I'm working on [sbx_subsample](https://github.com/louiejtaylor/sbx_subsample), an extension to subsample decontaminated reads. 

Basically, the rule makes a backup of the full decontaminated read files (into `QC_FP/decontam/full/`), before subsampling those backups and dropping the subsampled fastqs into `QC_FP/decontam/`. (We can talk about whether that's the right approach later.) The relevant parts of the rules are here:

```python
rule sample_reads:
    input:
        str(QC_FP/'decontam'/'{sample}_{rp}.fastq.gz')
    output:
        str(QC_FP/'decontam'/'full'/'{sample}_{rp}.fastq.gz')
    ...

rule all_subsample:
    input:
         expand(str(QC_FP/'decontam'/'full'/'{sample}_{rp}.fastq.gz'), sample=Samples.keys(), rp=Pairs)
```

I was getting two *strange* errors where it seemed as though random keywords from different rules were being slapped onto the sample names:

```
Building DAG of jobs...
InputFunctionException in line 18 of $SUNBEAM_DIR/rules/qc/qc.rules:
KeyError: 'SRR935340_hostreads'
Wildcards:
sample=SRR935340_hostreads
```

```
Collecting target genomes... done.
Building DAG of jobs...
InputFunctionException in line 18 of $SUNBEAM_DIR/rules/qc/qc.rules:
KeyError: 'full/SRR935339'
```

However, [a simple change](https://github.com/louiejtaylor/sbx_subsample/commit/97f026b4d39fc8c270c86c5b42586a4d46415265) fixed the problem--basically, it seems that the problem was somehow that I was working on a subfolder within the `decontam` directory. This works perfectly (although I don't see why it should be any different):

```python
rule sample_reads:
    input:
        str(QC_FP/'decontam'/'{sample}_{rp}.fastq.gz')
    output:
        str(QC_FP/'backup'/'{sample}_{rp}.fastq.gz')
    ...

rule all_subsample:
    input:
         expand(str(QC_FP/'backup'/'{sample}_{rp}.fastq.gz'), sample=Samples.keys(), rp=Pairs)
```

I don't know if this is a Sunbeam problem, a Snakemake problem, or a Louis-doesn't-know-Snakemake-quirks problem. And since I found a workaround, it's not imperative that we figure this out immediately, but I think this might really give someone a headache in the future (I sure am confused). Any thoughts?

**Comment by ressy** [comment]
I finally had a little time to have a look at this too, and I suspect it's a Snakemake issue.  I'm making a minimal working example in plain Snakemake that I'll put here when it's done.  (It looks like something subtly weird in how its pattern matching and lambda function feature interact, but nothing really Sunbeam-specific, I think.)

**Comment by ressy** [comment]
I take it back, it's on us, but it's a very subtle issue.  (Good catch!)

**tl;dr since this turned into a small essay: I think we should add a global constraint on possible values for the `sample` wildcard so it never tries to include directory names in it.**

Here's a single Snakefile that exemplifies the issue:

```
import pathlib
pathlib.Path("samplename_input.txt").touch()

SMPS = {"samplename": "samplename_input.txt"}

rule all_subdir:
    input:  "B/subdir/samplename.txt"

rule C:
    output: "B/subdir/{sample}.txt"
    input:  "B/{sample}.txt"

rule B:
    output: "B/{sample}.txt"
    input:  "{sample}_A.txt"

rule A:
    output: "{sample}_A.txt"
    input:  lambda wildcards: SMPS[wildcards.sample] # Case 1
    #input:  lambda wildcards: SMPS.get(wildcards.sample, "") # Case 2
    #input:  lambda wildcards: wildcards.sample + "_input.txt" # Case 3
```

The "Case 1" input for rule A is like what we currently have when we look up sample paths, and it fails here.

```
$ snakemake -s subdir.rules -nrp
Building DAG of jobs...
InputFunctionException in line 17 of /home/jesse/dev/sunbeam-185/subdir.rules:
KeyError: 'subdir/samplename'
Wildcards:
sample=subdir/samplename
```

The other two that are commented out actually work.

Three things that I think factor in:

* Snakemake works backwards from output toward input, resolving wildcards starting with the ordinary string paths of the final outputs.  So it needs to figure out which candidate wildcard-containing strings *could* be a match for these final outputs.
* Snakemake has no problem with including directory separators as parts of its wildcards (like sample="dir/subdir") so it considers that possibility alongside others.  (I think the order it uses to check these different cases might be random which would explain why you saw slightly different errors across runs.)
* When Snakemake is resolving the DAG, it'll call functions that define inputs for rules as part of that resolution, possibly multiple times for multiple candidate sets of wildcards.   It takes into account if candidate inputs paths don't exist, but still needs to call the function to get the string values.  This is the part that's our problem.

For our situation here, it works all the way back to the appropriate adapter_removal rule in qc.rules and gives candidate wildcards to the functions we use to look up file paths from the Samples dictionary.  Since some of those values don't exist in the keys for Samples, it throws KeyError, but it should really never be using those slash-containing values in the first place.

The [documentation on using a function as an input file](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#functions-as-input-files) says:
> it is best practice to make sure that it can properly handle all possible wildcard values your rule can have. In particular, input files should not be combined with very general rules that can be applied to create almost any file: Snakemake will try to apply the rule, and will report the exceptions of your input function as errors.

I don't think that's quite what happens (if I make the function return an empty string for non-matching keys, in Case 2 above, the graph resolution actually does work as expected), but point taken, we should be more restrictive in the possible wildcard values.

Possible fixes:

 * Add a constraint on the `sample` wildcard, so it won't attempt to include slashes inside the wildcard values
 * Just make the pattern strings more rule-specific so it doesn't have to work its way through similar-looking possibilities (I think that's your fix in sbx_subsample)

I like the first option since should protect us against similar problems down the line.  We can set that globally in the Snakefile and let it apply to any use of the `sample` wildcard.

```
wildcard_constraints:
  sample="[^/]+"
```

Does that seem reasonable?  I'll try the tests from scratch with that defined and make sure everything still resolves the same way.

**Comment by ressy** [comment]
(Oh and that change *does* allow you to use those original filename patterns and subdirectories for sbx_subsample-- I just realized I didn't explicitly go back and check that until just now.  )

**Comment by eclarke** [comment]
Thank you guys for figuring this out- that was a really neat debugging session to watch :D
